### DISCUSSION TITLE
"AccessDeniedAccess Denied" Error on images after changing Project on Github repos to public from private.

### DISCUSSION THREAD

**Discussion by LawJunHao** [community/community#56049] [Repositories]
### Select Topic Area

Bug

### Body

TLDR: 
Edit 19 Jun 2023: No solution has been found as of 19 June 2023, The only way is to find other apps and platforms to store images related to your project progress and not depend on github :(

Edit 24 July 2023: Still no solution to be found LOL. At this point I've given up on any possible fixes from github side and migrated the entire thing to another repository to salvage what I can. If you're a new comer here welcome! There is still no solution haha. 

Hello all, recently I've changed my project from private to public just for school submission related things. I'm the owner of the repos and project itself. Upon changing I checked the images inside projects and suddenly it gave me the error message "AccessDeniedAccess Denied" as seen below (I will censor a few things because i dont know what kind of error this is spitting out). The repos is fine however, nothing much changed. I panic and changed it back to private now thinking it was some permission error but it still seems to spit out the same thing. 

![image](https://github.com/community/community/assets/92876312/34f4916d-493a-4a30-b93f-01dfc713d9a2)

Images attached to project issues opened used to have a preview inside of them but now it's just replaced with an image icon as seen below in this image. 

![image](https://github.com/community/community/assets/92876312/fc9124ca-01a6-4a1a-8b94-5d31c80a5ab5)

What can i do to fix this permission error? it seems weird to have it because I've dug around the permission settings and I've set most of my team mates and myself to be admin of everything which still doesn't seem to work.

**Comment by SpiralBL0CK** [comment]
Did you find the solution ? Asking bc just happed to me rn :)

**Reply by LawJunHao** [reply]
Unfortunately... no. But I've made enough screenshot archives to fill in the gaps missing due to this error. The solution is to make sure keep project always in a certain permission from the very start I guess :(

**Reply by SpiralBL0CK** [reply]
i guess so the way i solved my issue was by importing whatever i was doing in gitbook and it re-uploading each screenshot :/ so i guess it's exactly as u said from very start have certain permissions

**Comment by SomiaNasir** [comment]
Are you able to solve the problem?  
I am having the same issue after making the repo public. There is a difference in links generated between this repo and previous ones I had. This one has something like https ://github.com/user_name/repo_name/assets/123/123456-abc  
Previous ones has like https ://user-image.githubusercontent.com/123/123456-abc.png and they did not had this problem of access denied after making them public. Could it be the reason?

**Reply by LawJunHao** [reply]
That's interesting that you pointed out the links used for the images appear to be different. I highly suspect when GitHub changes repos from private to public they change the asset names as well. Which could be the reason why all our previous image assets linked to the uploaded image could be invalid. But when we change back it doesn't use back the old link. But so far until now I haven't found a solution to this...

**Comment by ManoloTonto1** [comment]
Same here, this just shows up
![image](https://github.com/community/community/assets/74105210/5e3f86be-4669-4f0a-8c0d-861b3281f338)

**Reply by LawJunHao** [reply]
Exactly the same issue I'm facing, welp all we can do right now is suck it up for a bit and raise the awareness as much as we can to this community channel. Maybe we can get an answer from the mods or creators at some point.

**Comment by DanielDW23** [comment]
Same problem in some of my repositories

**Comment by NinDevs** [comment]
Same issue here

**Comment by kplr-io** [comment]
Same here :(((((

**Reply by NinDevs** [reply]
> Same here :(((((

I fixed it by just using Imgur tbh lol

**Reply by LawJunHao** [reply]
Thats a pretty good solution too, I myself started taking note in other tools that help w this kind of stuff (note taking apps) like notion.

**Comment by GHomeSmartMiniSmartPlug** [comment]
When encountering an "**[Access Denied](https://seller.hewglo.com/boost-your-business-10-smart-home-devices-for-sellers-on-hewglo)**" error on images after changing a project on GitHub from private to public, it is likely due to the permissions settings not being updated properly. GitHub's access controls restrict access to certain resources based on the project's visibility settings. When changing a project from private to public, ensure that the image files and their corresponding directories have the correct permissions for public access. Double-check the file permissions, including read and write permissions, and ensure they are set to allow public viewing. Additionally, confirm that the image files are in the correct location and properly linked within the project.

**Reply by michalis0** [reply]
When you copy-paste images in github they get pushed to some s3 amazon bucket, eg ...s3.amazonaws.com/28807066/... for which you cannot change the permissions afterwards. Very weird bug github! Fix it!

**Comment by RalphBacon** [comment]
I've just stumbled across this bug. Steps to reproduce:

I created a repository, made it private and uploaded all my text and copy/pasted images just as I have done several dozen times before.

Today I made the repository public, which it became, except all the images were "Access Denied" both in the repository and even individually, for me (when I copied the URL for an image into a new web browser tab).

A typical example of an [image URL is this](https://private-user-images.githubusercontent.com/20911308/250829148-13140b0e-9427-46c0-8f3a-d5cc77669eb7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJrZXkiOiJrZXkxIiwiZXhwIjoxNjg4NjU3ODY5LCJuYmYiOjE2ODg2NTc1NjksInBhdGgiOiIvMjA5MTEzMDgvMjUwODI5MTQ4LTEzMTQwYjBlLTk0MjctNDZjMC04ZjNhLWQ1Y2M3NzY2OWViNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMwNzA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMDcwNlQxNTMyNDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zNDM3NTRkYTA4NDQ5MWZkMzA3MjdiZmE1MGI0NDRmZDk2NjE2OTEzZDE4NGMzMmY3YjFmZmM2NTIwNjY0MWE5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.S7UqeCsO5-SYHCJRqDaDDtrKBw6TEa9DM_R7M05SRgs)

Making the repository PRIVATE again, showed all the images once again; but making it PUBLIC once more ensured all the images disappeared again. So I edited the (now PUBLIC) repository and replaced all the images (dragged them in) and it's now as it should be.

This is a ridiculous error that must have appeared in June 2023 (everything was fine before that). I don't understand why this wasn't discovered and fixed immediately.

Does this mean that temporarily PRIVATE repositories are now useless if they contain images?

**Reply by arnaldmarcer** [reply]
We do have the same problem. Just made a repository public and lost all images images inserted in the issues. The surprising thing is that we did not lose all of them, just for the more recent issues. I do not want to make the repository private again and do not know how to recover those explanatory images we had in the issues. Anybody found a solution?

**Comment by MohamedWageh09** [comment]
I'm facing same problem too, that's very annoying idk why they didn't fix it yet :(

**Comment by yuhang559** [comment]
I'm facing the same problem, so I had to make it private again :(

**Comment by WeaverDev** [comment]
Same issue here

**Comment by JacobH140** [comment]
I have the same issue, I suppose the strategy for now is to manually re-upload images after changing visibility?

**Reply by RalphBacon** [reply]
Well, that's the workaround, and I can do that for 3 images but not for 67 images. 

I'm stunned and disappointed that such a major bug hasn't yet been addressed.

**Comment by taofikkhris** [comment]
The "AccessDeniedAccess Denied" error on images after changing a GitHub repository from private to public is likely due to cached permissions or access controls that have not yet fully propagated across GitHub's servers and CDN (Content Delivery Network). When you change a repository's visibility from private to public, it might take some time for the changes to be fully applied.

Here are some steps you can follow to troubleshoot and resolve the issue:

Clear Your Browser Cache: Sometimes, the browser cache may still have the old access control settings stored. Clear your browser cache and try accessing the images again.

Wait for Propagation: After changing the repository's visibility to public, wait for some time (usually a few minutes) to allow GitHub's servers and CDN to propagate the changes.

Check Image URLs: Ensure that the URLs of the images you are trying to access are correct. Public images in a GitHub repository should be accessible via their raw URLs, like https://raw.githubusercontent.com/username/repo-name/master/path/to/image.png.

Verify Permissions: Double-check that the repository is indeed public and that the images have the necessary permissions for public access. You can go to the repository settings and check the "Danger Zone" section to verify the visibility settings.

Check File Paths: Make sure the file paths to the images are correct and have not been modified during the repository's transition from private to public.

Verify Image Uploads: If the images were uploaded recently after changing the repository to public, ensure that the image uploads were successful, and the images are present in the repository.

Test from Another Device/Network: Try accessing the images from another device or network to ensure the issue is not specific to your current setup.

Contact GitHub Support: If the issue persists even after trying the above steps, it's best to reach out to GitHub support for further assistance.

Remember that changes to repository settings, especially visibility changes, may take some time to reflect across GitHub's systems. If you've followed the steps above and are still experiencing the issue, waiting a bit longer might resolve the problem as the changes continue to propagate.

**Reply by michalis0** [reply]
Your answer is not valid. This problem appears when copy pasting images in github so the original urls are some links in aws and the permissions are not reset.

**Reply by RalphBacon** [reply]
@michalis0 I concur.

In my private GitHub the pasted image urls look like this:
`https://private-user-images.githubusercontent.com/20911308/250829148-13140b0e-9427-46c0-8f3.... (long string)`

The problem is that this url is not made non-private when making the GitHub public and everyone (including the creator) now gets a 404.

As reported on July 6th. Still. No. Resolution.

**Reply by LawJunHao** [reply]
Thats quite a useful and valid lists of things to check for new people to this thread! However I think I've tried most of them when first posting this issue back in may haha. At this point in august I've already fully given up on that repository and it's contents and migrated to another platform to store the project progress. But if yall find any more solutions do post here, seeing how people still interact with this means the issue is still on going.

**Comment by richinatl** [comment]
I just came across this issue. I made a repository directly on Visual Studio and VS made it private. I realized it, made it public, then had this issue. I took a screenshot of the screenshot to change the file name and location and so far that worked. What a weird pain

**Comment by Sahill17** [comment]
I'm facing the same issue, had to upload all the images in public repository again.

**Reply by RalphBacon** [reply]
Unfortunately, no one at GitHub HQ seems to care about this bug. Perhaps they've all been seconded to the Windows 12 Dev team. 🤷‍♂️

**Reply by Mickmlone44** [reply]
Hello

**Comment by yutotakano** [comment]
Not only is the problem for all upload images embedded from within READMEs and other files, it also applies to all uploaded images referred to in GitHub Releases -- I just had all my screenshot images for all past releases go Access Denied when I made the repository public, and had to go through each one and re-upload them (if I remembered which images they were). The fact that this bug thread seems to have had little attention from GitHub is quite frustrating and disappointing, I hope it gets fixed soon.

**Comment by avkrishnamurthy** [comment]
Same issue. Don't understand how this hasn't been addressed yet.

**Comment by spendyaala** [comment]
SAME ISSUE HERE TOO. WILL THIS EVER GET RESOLVED?

**Reply by RalphBacon** [reply]
Not until it affects someone important at M$ HQ.

**Comment by sergiosolorzano** [comment]
Same issue here. Behaviour is clearly outlined in this post. Make repo public from private and access denied error breaks all images. Can be solved by going back to private setting, save all videos/images, then back to public, re-upload. Very painful. I've raised github support ticket. Github team please help!

**Comment by MoatazBadawy** [comment]
IDK How GitHub team doesn't respond to these people! disappointed!

**Comment by dipanshuchaubey** [comment]
The only solution that worked for me was to re-upload the assets after making your repository public. You will notice that when your repository was private the asset URL looked something like this `https://github.com/your-user-name/your-repo-name/assets/your-user-id/a-random-uuid`.

But on a public repo, when a new asset is uploaded the URL looks like this `https://github-production-user-asset-6210df.s3.amazonaws.com/your-user-id/a-random-uuid.png`

It seems GitHub uses an S3 bucket for public repos as they expect resources from these repos to be accessed much more frequently than the private ones.

Hope it helps.

**Comment by RalphBacon** [comment]
Yes, it helps, but I don't understand the logic of what GitHub is doing. 

Why are the images not uploaded to their "final" (albeit pubic) storage in AWS as they cannot really be accessed, not even by their owner without knowing the exact URL (with that random uuid number). 

That way this entire problem would be solved as the images would not have to be copied from one location to another (assuming that is what is happening).

Seems to me that GitHub have dropped the ball on this one and are not picking it up. I'm amazed that millions are not complaining 🤷

**Reply by LawJunHao** [reply]
I dont think this issue is that well known because not everyone who uses github uses all these features as well. We just have to wait and see if any github staff spots this thread by chance and raises this issue to their dev team.

**Comment by amaimus** [comment]
Same problem here ❌❌

**Comment by amaimus** [comment]
Guys, I'm trying this solution and it seems to work: : https://stackoverflow.com/questions/14494747/how-to-add-images-to-readme-md-on-github

You just need to add the image to the repository and then link directly to the raw version of the image


`![alt text](https://github.com/[username]/[reponame]/blob/[branch]/image.jpg?raw=true)`

I'm using it to show a preview of a project I made and it works fine: https://github.com/amaimus/react-todo-app-typescript

![image](https://github.com/community/community/assets/35699916/82c6fbc0-3e49-40ba-a433-2a8e52b4da6a)

**Reply by arnaldmarcer** [reply]
Does not work for me.

**Reply by amaimus** [reply]
- Did you check that the image has a file type supported by github?

`gif, jpeg, jpg, mov, mp4, png, svg, webm`

pd: Even though it is supported by Github, the gifs don't work for me

- Did you place the image at the root of your repository?

**Reply by arnaldmarcer** [reply]
Yes, it is a png file. It became inaccessible after changing the repo from private to public.

The link within the issue is: 
<img width="528" alt="image" src="https://github.com/aescobarr/GeoPick/assets/462827/b9da595a-31f4-4532-b2c7-538749f33e4f">

I changed it to:
![image](https://github.com/aescobarr/GeoPick/assets/462827/b9da595a-31f4-4532-b2c7-538749f33e4f)

but it does not work.

**Reply by amaimus** [reply]
In one of your repositories I found a MD syntax that I haven't seen before but it seems to work as well

Try this:

`<img src='pathToYourImageFromRoot.png'/>`

Supposing that is `my-image.png` and it is in the `public` folder

`<img src='public/my-image.png'/>`


Notice that you can also link to URLs
Link: https://www.google.es/images/branding/googlelogo/1x/googlelogo_color_272x92dp.png
MD:
`<img src='https://www.google.es/images/branding/googlelogo/1x/googlelogo_color_272x92dp.png'/>`
Output:
<img src='https://www.google.es/images/branding/googlelogo/1x/googlelogo_color_272x92dp.png'/>

**Reply by arnaldmarcer** [reply]
My image from root is already the link I have. I do not have a public folder either. I created the image when I created the issue by pasting it. It got lost when going from private to public.

**Reply by amaimus** [reply]
Oh I understand, it also happened to me with a couple of repositories. 🫠

What I will do from now on is always save the image in the project itself, this way I prevent Github from losing them 😂

👀👀 Let me know if the previous solution worked for you

**Comment by AnanyaJha** [comment]
Facing the exact same issue with images included in reported GH Issues. We recently transitioned https://github.com/forcedotcom/Einstein-GPT-for-Developers from a private to a public repository. Since the transition, all previous images shared as part of an Issue AND ALL new images shared within our GH Issues are unavailable to us. Clicking on an image opens up a new page where we are presented with an AccessDenied error.

This problem is making triaging our open GH Issues extremely difficult.

**Comment by Syndesi** [comment]
I opened a GitHub support ticket, and got the following response:

> Hi Syndesi,
>  
> Thanks for reaching out to GitHub Support!
>  
> Sorry that you're having trouble with the attached images in your repository - this is an issue that's been identified as result of [delivering more secure private attachments](https://github.blog/changelog/2023-05-09-more-secure-private-attachments/).
> 
> When a private repository is changed to public, the images remain protected and cannot be loaded, even by authenticated users. You should be able to retrieve them by changing the repository back to private. If you'd like the images to be accessible publicly, the workaround is to reupload them whilst the repository is set to public.
>  
> I'm sorry about the inconvenience!
>  
> Regards,
> [name of the employee].

**Reply by Karmalakas** [reply]
> reupload them whilst the repository is set to public

Doesn't work...

**Reply by Syndesi** [reply]
@Karmalakas please reach out to the GitHub Support team with more information :)

**Reply by Karmalakas** [reply]
Can't find how to. It just ends up in useless help pages 🤷‍♂️

**Reply by Syndesi** [reply]
Direct link to create a ticket: https://support.github.com/contact?source=subtitle&tags=rr-general-technical

Requirements to create a ticket:
> If your account uses a paid GitHub product or you are a member of an organization that uses a paid product, you can directly contact GitHub Support.
> If your account uses GitHub Free, you can speak to GitHub users and staff on the [GitHub Community discussions](https://github.com/orgs/community/discussions) for most issues, and you can contact GitHub Support to report account, security, and abuse issues.

See also https://docs.github.com/en/support/contacting-github-support/creating-a-support-ticket

**Reply by Karmalakas** [reply]
So that basically says I can't contact support 🙂 Thanks

**Reply by ghufran2508** [reply]
Hi @Karmalakas .
So you want to set you want to set your image assets to public, I can help you if you want.

**Comment by RalphBacon** [comment]
I am very disappointed at GitHub's / Microsoft's response. 

I create PRIVATE repositories so I can work on them without others seeing them (ands commenting on them) before the project is finished, or before it's been authorised to go public. It is, after all, a Work In Progress,

Then I release the entire, completed and verified project as a PUBLIC project, including all the carefully crafted images (part of the documentation) that may have taken weeks to gather, so they can be publicly viewed.

The workaround seems to NOT drag-and-drop images into pages; instead, save all the images into a folder and refer to them specifically, by name, in the text. It's more work, but what I used to do before I discovered the drag-and-drop method. 

I guess I will have to try this out to make sure it still works for a project that is created as a PRIVATE project that then becomes PUBLIC. 😓

**Reply by rickstaa** [reply]
Shootout to @Syndesi for addressing this issue! 

@RalphBacon, I agree with you that the current behavoir could be better. Maybe the GitHub team could consider incorporating an `also make attachments public` checkbox adjacent to the `Change repository visibility` option? This feature would significantly enhance user convenience by allowing them to publicize attachments selectively, maintaining the option for secure handling of private attachments without compromising the confidentiality of older attachments 🤔.

**Reply by RalphBacon** [reply]
You can ask them, but given this "issue" has existed for about 5 months now, I'm assuming they won't change their stance. What's perhaps most galling is that MS changed GitHub's behaviour overnight; no warning or even mention that this was going to happen. One of my users brought it to my attention after a release. Awkward. 🤷

I just tried out my own suggestion of reverting back to using an "images" folder and that seems to be fine. Everything created whilst the project was PRIVATE, made it PUBLIC and it still shows the image (as I would expect, after all, it's a file in a folder).

**Comment by ghufran2508** [comment]
Hi All,

I hope this message finds you well! 🌟
I noticed you recently encountered an issue with image access in your README.md after making your repository public. I've experienced a similar situation before, and I wanted to share a solution that worked for me.

It seems like the access denied error might be related to the image links not updating properly when the repository status changes from private to public. To resolve this, I recommend re-uploading the images by dragging and dropping them into your README.md and updating only the image links where necessary. This time github mark this asset as public and the access denied problem will no longer be there.

I recently faced a similar challenge, and by following these steps, I was able to resolve the access denied issue.
